### ISSUE TITLE
Document API

### ISSUE THREAD

**Issue by olorin** [anchor/sieste#19]
There's been some confusion over what expected behaviour is/whether given query results are buggy or not. We should document the expected behaviour of each form of query, including examples of what will and will not match. In addition, document things like default page sizes and how to change them.

(CC @certifiedwaif)

**Comment by olorin** [comment]
Also, RFC: should we consider API versioning?

@glasnt

**Comment by glasnt** [comment]
Yes, we need more documentation than just the readme and inline comments. 
And we have a kinda versioning, but it's very out of line with current best practices. See /releases vs sieste.cabal
View to include both in the 0.5(minor bump) lucene branch

**Comment by certifiedwaif** [comment]
I was expecting that the form of query results for search would be the same whether the search terms changed or not, and that every line returned from search would have the same fields.

**Comment by glasnt** [comment]
> I was expecting that the form of query results for search would be the same whether the search terms changed or not

The results format should always be an array of sourcedicts in a JSON parsable format. Should.

> and that every line returned from search would have the same fields.
> Ideally for any one origin, all input would be the same, therefore all fields would be the same, but this is not validated by the vault, nor by sieste (as far as I am aware. Horray document storage). Unsure on the viability of returning nils for empty fields on records that do not contain the full set of known tags for a origin.

**Comment by olorin** [comment]
> The results format should always be an array of sourcedicts in a JSON parsable format. Should.

As far as I'm aware this is the case. Can you clarify what you mean by 'form' @certifiedwaif?

> and that every line returned from search would have the same fields.

This isn't something sieste can enforce (maybe chevalier could, but it'd be a big change - maybe we could discuss that possibility next week?) - at the moment it's the collectors' responsibility, and not assumed to be guaranteed in the general case.

One easy change we can make to ameliorate this is to have chevalier discard old sourcedicts before indexing new ones - at the moment it just performs a merge with the SD it gets from the contents daemon, i.e., it adds any fields present in the contents list and not Elasticsearch, but it doesn't change fields which are in Elasticsearch (and therefore indexed previously) but are not in the current contents list. I'll create an issue for that now.

**Comment by olorin** [comment]
anchor/chevalier#25

**Comment by certifiedwaif** [comment]
The same keys. But I suspect that's not possible. It would be good to at least know what keys I could get in each of the source dicts.
